controller/backing_image_controller.go

```diff
@@ -596,28 +596,121 @@ func (bic *BackingImageController) syncBackingImageFileInfo(bi *longhorn.Backing
 	return nil
 }
 
-func (bic *BackingImageController) updateDiskLastReferenceMap(backingImage *longhorn.BackingImage) error {
-	replicas, err := bic.ds.ListReplicasByBackingImage(backingImage.Name)
+func (bic *BackingImageController) updateDiskLastReferenceMap(bi *longhorn.BackingImage) error {
+	log := getLoggerForBackingImage(bic.logger, bi)
+
+	bids, err := bic.ds.GetBackingImageDataSource(bi.Name)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+		log.Warn("Cannot find backing image data source, but controller will ignore it and continue updating disk last reference map")
+	}
+
+	replicas, err := bic.ds.ListReplicasByBackingImage(bi.Name)
 	if err != nil {
 		return err
 	}
-	disksInUse := map[string]struct{}{}
+	filesInUse := map[string]struct{}{}
 	for _, replica := range replicas {
-		disksInUse[replica.Spec.DiskID] = struct{}{}
-		delete(backingImage.Status.DiskLastRefAtMap, replica.Spec.DiskID)
+		filesInUse[replica.Spec.DiskID] = struct{}{}
+		delete(bi.Status.DiskLastRefAtMap, replica.Spec.DiskID)
 	}
-	for diskUUID := range backingImage.Status.DiskLastRefAtMap {
-		if _, exists := backingImage.Spec.Disks[diskUUID]; !exists {
-			delete(backingImage.Status.DiskLastRefAtMap, diskUUID)
+	for diskUUID := range bi.Status.DiskLastRefAtMap {
+		if _, exists := bi.Spec.Disks[diskUUID]; !exists {
+			delete(bi.Status.DiskLastRefAtMap, diskUUID)
 		}
 	}
-	for diskUUID := range backingImage.Spec.Disks {
-		_, isActiveDisk := disksInUse[diskUUID]
-		_, isRecordedHistoricDisk := backingImage.Status.DiskLastRefAtMap[diskUUID]
-		if !isActiveDisk && !isRecordedHistoricDisk {
-			backingImage.Status.DiskLastRefAtMap[diskUUID] = util.Now()
+	// Set timestamp for all candidates based on the replica usage first.
+	for diskUUID := range bi.Spec.Disks {
+		if _, exists := bi.Status.DiskFileStatusMap[diskUUID]; !exists {
+			continue
+		}
+		_, isActiveFile := filesInUse[diskUUID]
+		_, isRecordedHistoricFile := bi.Status.DiskLastRefAtMap[diskUUID]
+		isFirstFile := bids != nil && !bids.Spec.Started && diskUUID == bids.Spec.DiskUUID
+		if !isActiveFile && !isRecordedHistoricFile && !isFirstFile {
+			// The timestamp may be removed later based on the HA requirement.
+			bi.Status.DiskLastRefAtMap[diskUUID] = util.Now()
 		}
 	}
+
+	// Determine the final cleanup entries based on HA requirement as well as
+	// the current file state map.
+	// TODO: Make `haBackingImageCount` configure when introducing HA backing image feature
+	haBackingImageCount := 1
```

Comment:
Move the HA check to the cleanup function rather than this timestamp setting function